### ISSUE TITLE
WhatFailureGroupHandler not handling bubbling.

### ISSUE THREAD

**Issue by adjenks** [Seldaek/monolog#946]
Bubbling doesn't appear to be working, I tried changing the order of handlers in the array passed to the WhatFailureGroupHandler, but it doesn't seem to affect it, I also tried setting bubble to true and false. No combination of settings seem to prevent bubbling. I also tried setting bubbling on the actual WhatFailureGroupHandler in addition to the handlers it contains. How do I stop bubbling with this handler?

**Comment by Seldaek** [comment]
As far as I can see in the code it does the right thing, so not sure what you mean here. If you set bubble to false in the constructor, and any of its handlers is handling a given message, it should return true in handle() which stops the bubbling chain.

I'll need more details if you want help on this one I think.

**Comment by adjenks** [comment]
If you set bubble to false on sub-handler in the WhatFailureGroupHandler array, they do not trigger bubbling to stop.
This is because WhatFailureGroupHandler simply iterates over the handers and doesn't check their bubble status, it simply calls all of them like so:
```
        foreach ($this->handlers as $handler) {
            try {
                $handler->handle($record);
            } catch (\Throwable $e) {
                // What failure?
            }
        }
```

After discovering this, I was going to modify WhatFailureGroupHandler to do something like this:

```
        foreach ($this->handlers as $handler) {
            try {
                if(true == $handler->handle($record)){
                    break;
                }
            } catch (\Throwable $e) {
                // What failure?
            }
        }
```

Similar to how bubbling is controlled in Logger. However I realized that I would rather just be able to let Logger not die, since logging is important and WhatFailureGroupHandler seemed redundant, which is why I made that fork and pull request.

So as you can see above, WhatFailureGroupHandler simply calls all of its handlers, and ignores bubbling.
It is true that the WhatFailureGroupHandler does obey bubbling itself, if it's own bubble is set to false, it will stop any logging if there are other handlers in the log registry, however, internally it just ignores bubbling.

**Comment by adjenks** [comment]
Here's an example:
```
<?php

use Monolog\Logger;
use Monolog\Handler\StreamHandler;
use Monolog\Handler\WhatFailureGroupHandler;
require 'vendor/autoload.php';


// create a log channel
$log = new Logger('name');
$filehandler1 = new StreamHandler($_SERVER["DOCUMENT_ROOT"].'/', Logger::DEBUG,true); //fails without file name, should continue on after failure
$filehandler2 = new StreamHandler($_SERVER["DOCUMENT_ROOT"].'/logs/testlog.log', Logger::DEBUG,false); //should stop bubbling here, bubble set to false
$filehandler3 = new StreamHandler($_SERVER["DOCUMENT_ROOT"].'/logs/testlog2.log', Logger::DEBUG,true); //shouldn't write here, but it does

$wfgh = new WhatFailureGroupHandler([$filehandler1,$filehandler2,$filehandler3]); //it fails at 1, succeeds at 2, ignores false bubble, proceeds to 3
$log->pushHandler($wfgh);

// add records to the log
$log->warning('bar');

```

**Comment by Seldaek** [comment]
The reason is that this handler is a Group handler, and as such it means anything going into it is dispatcher to all items of the group..

If you want to control bubbling more granularily, I'd suggest you set the bubble on the whatfailuregrouphandlers, and wrap every handler you have into its own whatfailuregrouphandlers instance.

**Comment by adjenks** [comment]
Right, that should work. Perhaps I'll do that. It seems like a bit of a workaround though.
Also, it wouldn't be able to report any failures in the handler because WhatFailure doesn't do that yet, which is part of the reason why I rolled my own version, I wanted to know when my loggers were failing and which ones. I noticed it was a feature request here [https://github.com/Seldaek/monolog/issues/860](https://github.com/Seldaek/monolog/issues/860)
Perhaps I could help implement that? I will do what you suggested and 

> pass it in the constructor, or allow passing a set of handlers that are "safer" and less likely to fail and should be used for logging handler failures?

Mind if I shoot you a pull request for that?

**Comment by Seldaek** [comment]
No a PR for that sounds good. Please target the 1.x branch though not 
master.

A set of fallback handlers to report logging errors sounds good.

**Comment by adjenks** [comment]
Actually wait, that doesn't quite solve my problem. Since I wanted to use bubbling to control loggers as fallback loggers.
Example:
```
 Try to log to A,
 if A fails ignore the exception and log to B
 if A succeeds, check it's bubble status
   if bubble status is true log to B
   if bubble status is false end
```

With WhatFailure you can ignore errors, and you can tell it to stop bubbling, but you can't let it continue to a fallback if there was an error in A, and A was wrapped in a WhatFailureGroupHandler. I think I already went through this train of though before. I tried wrapping them all in WhatFailureGroupHandlers. But I couldn't get the granularity of flow that I wanted.

Basically I wanted to :
log to a database, if the database fails, log to a file, and also report the database log failure.
I couldn't achieve this with wrapping them all.

**Comment by Seldaek** [comment]
Yeah that sounds like a custom handler implementing exactly that logic 
would be the best option though. Creating handlers is pretty easy, 
monolog doesn't have to provide every possible handler out there :)

**Comment by adjenks** [comment]
Good point it is easy. I just thought that either the group handler or the primary logging implementation should handle internal failures gracefully. That was actually my first choice, I was going to write a custom one, then I thought the WhatFailureGroupHandler might as well do it, so I started implementing it all there, then I thought, why put it there, it should be built right in.
I'll just make a custom one though.

**Comment by Seldaek** [comment]
Closing as https://github.com/Seldaek/monolog/pull/1012 allows you to do more advanced error handling.
